components/omega/doc/design/PGrad.md

```diff
@@ -0,0 +1,196 @@
+Therefore the centered pressure gradient will be calculated as:
+
+$$ T^p_k = \frac{1}{d_e} \left( \widehat{v}_{k,e} \sum_{i \in CE(e)} n_{e,i} \overline{p}_{k,i} + \sum_{i\in CE(e)} n_{e,i} \overline{\phi}_{k,i}\right), $$
+
+$$ = \frac{1}{d_e} \left(  \sum_{i \in CE(e)} n_{e,i} \left( \widehat{v}_{k,e}\overline{p}_{k,i} + \overline{\phi}_{k,i} \right) \right), $$
+
+with the vertical averaging operator defined as:
+
+$$ \overline{(\cdot)} = \frac{1}{2} \left((\cdot)_{k+1/2} + (\cdot)_{k-1/2} \right)$$
+
+and the horizontal averaging operator:
+
+$$\widehat{(\cdot)} = \frac{1}{2} \left( (\cdot)_{i=1} + (\cdot)_{i=2}\right)$$
+
+## 4 Design
+
+### 4.1 Data types and parameters
+
+The `PressureGradient` class will be used to perform the horizontal gradients of the pressure and geopotential
+```c++
+class PressureGrad{
+    public:
+    private:
+        PressureGradCentered CenteredPGrad;
+        PressureGradHighOrder HighOrderPGrad;
+        PressureGradType PressureGradChoice;
+        I4 NVertLevels;
+        I4 NChuncks;
+        Array2DI4 CellsOnEdge;
+        Array1DReal DvEdge;
+        Array1DReal EdgeSignOnCell;
+}
+```
+The user will select a pressure gradient option at runtime in the input yaml file under the pressure gradient section:
+```yaml
+    PressureGrad:
+       PressureGradType: 'centered'
+```
+An `enum class` will be used to specify options for the pressure gradient used for an Omega simulation:
+```c++
+enum class PressureGradType{
+   Centered,
+   HighOrder
+}
+```
+The functions to compute the centered and high order pressure gradient terms will be implemented as functors and the pressure gradient class will have private instances of these classes.
+### 4.2 Methods
+
+```c++
+class PressureGrad{
+    public:
+        static PressureGrad *create();
+        void computePressureGrad();
+        void computePressureGradBtr();
+    private:
+
+}
+```
+#### 4.2.1 Creation
+
+The constructor will be responsible for storing any static mesh information as private variables and handling the selection of the user-specified pressure gradient option.
+```c++
+PressureGrad::PressureGrad(const HorzMesh *Mesh, int NVertLevels, Config *Options);
+```
+
+The create method will take the same arguments as the constructor plus a name.
+It calls the constructor to create a new pressure gradient instance, and put it in the static map of all pressure gradients.
+It will return a pointer to the newly created object.
+```c++
+PressureGrad *PressureGrad::create(const std::string &Name, const HorzMesh *Mesh, int NVertLevels, Config *Options);
+```
+
+#### 4.2.2 Initialization
+
+The init method will create the default pressure gradient and return an error code:
+```c++
+int PressureGrad::init();
+```
+
+#### 4.2.3 Retrieval
+
+There will be methods for getting the default and non-default pressure gradient instances:
+```c++
+PressureGrad *PressureGrad::getDefault();
+PressureGrad *PressureGrad::get(const std::string &Name);
```

Comment:
What is the use case for multiple gradient instances?

Reply:
My thought was that we could use a separate instance for the baroclinic and barotropic pressure gradients. But perhaps this doesn't make much sense to do that.

Reply:
Also, I've just been using this as a template since it was established early on for a lot of the infrastructure classes (when it made sense to support multiple instances). Now that may not be needed for Omega-1, is there a convention we want to use to do the retrieval for a single instance between the `init` and `run` routines? This is something we were also wondering for the EOS (@alicebarthel)

Reply:
It's just like the multiple instance - you just keep the managed (unique) pointer to the single instance rather than saving a map of multiple instances (see for example the single calendar in TimeMgr). And it's fine to have a single instance of the baroclinic and a single instance of barotropic with a get for each.  But a retrieval by name involves a string comparison and search through the map so adds some unnecessary overhead if we don't need it. It's one of the things that slowed down MPAS.